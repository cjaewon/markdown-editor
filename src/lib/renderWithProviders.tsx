import React from 'react';
import { MockedProvider, MockedResponse } from '@apollo/react-testing';
import { render } from '@testing-library/react';
import { createStore } from 'redux';
import { Provider } from 'react-redux';
import rootReducer, { RootState } from '../modules';
import { createMemoryHistory, MemoryHistory } from 'history';
import { Router } from 'react-router-dom';
import { safe } from './utils';

type Options = {
  initialState?: RootState;
  mocks?: MockedResponse[];
  routeOptions?: {
    route: string;
    history?: MemoryHistory;
  };
};

export default function renderWithProviders(
  ui: React.ReactNode,
  { initialState, mocks, routeOptions }: Options = {},
) {
  const store = createStore(rootReducer, initialState);
  const history =
    safe(() => routeOptions!.history) ||
    createMemoryHistory({
      initialEntries: [safe(() => routeOptions!.route) || '/'],
    });

  return {
    ...render(
      <Router history={history}>
        <MockedProvider mocks={mocks} addTypename={false}>
          <Provider store={store}>{ui}</Provider>
        </MockedProvider>
      </Router>,
    ),
    store,
    history,
  };
}
