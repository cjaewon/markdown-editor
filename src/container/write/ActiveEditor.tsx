import React, { useEffect, useState, useCallback, useRef } from 'react';
import { useSelector, useDispatch, shallowEqual } from 'react-redux';
import { RootState } from '../../modules';
import {
  clearEditor,
  prepareEdit,
  setInitialBody,
  setInitialTitle,
} from '../../modules/write';
import EditorPanesContainer from './EditorPanesContainer';
// import QuillEditorContainer from './QuillEditorContainer';
import { useLocation } from 'react-router';
import qs from 'qs';
import { useQuery } from '@apollo/react-hooks';
import {
  READ_POST_FOR_EDIT,
  ReadPostForEditResponse,
  GET_LAST_POST_HISTORY,
  GetLastPostHistoryResult,
} from '../../lib/graphql/post';
import { safe } from '../../lib/utils';
import PopupOKCancel from '../../components/common/PopupOKCancel';

export type ActiveEditorProps = {};

const ActiveEditor: React.FC<ActiveEditorProps> = () => {
  const [newPost, setNewPost] = useState(false);
  // const mode = useSelector((state: RootState) => state.write.mode);
  const postId = useSelector((state: RootState) => state.write.postId);
  const [askLoadTemp, setAskLoadTemp] = useState(false);
  const initialized = useRef(false);

  const dispatch = useDispatch();
  const location = useLocation();

  const { id }: { id?: string } = qs.parse(location.search, {
    ignoreQueryPrefix: true,
  });

  useEffect(() => {
    if (!id) {
      setNewPost(true);
    }
  }, [id]);

  const readPostForEdit = useQuery<ReadPostForEditResponse>(
    READ_POST_FOR_EDIT,
    {
      variables: {
        id,
      },
      skip: !id || newPost, // skip if id is missing
    },
  );

  const getLastPostHistory = useQuery<GetLastPostHistoryResult>(
    GET_LAST_POST_HISTORY,
    {
      variables: {
        post_id: id,
      },
      skip: !id,
      fetchPolicy: 'network-only',
    },
  );

  useEffect(() => {
    return () => {
      dispatch(clearEditor());
    };
  }, [dispatch]);

  const post = safe(() => readPostForEdit.data!.post);
  useEffect(() => {
    if (!post) return;
    if (initialized.current) return;
    dispatch(
      prepareEdit({
        id: post.id,
        title: post.title,
        body: post.body,
        tags: post.tags,
        description: post.short_description,
        urlSlug: post.url_slug,
        series: post.series,
        isPrivate: post.is_private,
        isMarkdown: post.is_markdown,
        isTemp: post.is_temp,
        thumbnail: post.thumbnail,
      }),
    );
    dispatch(setInitialBody(post.body));
    initialized.current = true;
  }, [dispatch, post]);

  const lastPostHistory = safe(() => getLastPostHistory.data!.lastPostHistory);

  const onApplyTempSave = useCallback(() => {
    if (!lastPostHistory) return;
    dispatch(setInitialTitle(lastPostHistory.title));
    dispatch(setInitialBody(lastPostHistory.body));
    setAskLoadTemp(false);
  }, [dispatch, lastPostHistory]);

  useEffect(() => {
    if (!lastPostHistory) return;
    if (!post) return;

    const equals = shallowEqual(
      {
        title: lastPostHistory.title,
        body: lastPostHistory.body,
      },
      {
        title: post.title,
        body: post.body,
      },
    );
    if (equals) return;
    if (new Date(post.updated_at) < new Date(lastPostHistory.created_at)) {
      setAskLoadTemp(true);
    }
    // if (lastPostHistory.is_markdown !== post.is_markdown) {
    //   dispatch(convertEditorMode());
    // }

    // dispatch(changeTitle(lastPostHistory.title));
    // dispatch(setInitialBody(lastPostHistory.body));
  }, [dispatch, lastPostHistory, post]);

  if (id && !post && !postId) return null;

  return (
    <>
      <EditorPanesContainer />
      <PopupOKCancel
        visible={askLoadTemp}
        title="임시 포스트 불러오기"
        onCancel={() => setAskLoadTemp(false)}
        onConfirm={onApplyTempSave}
      >
        임시저장된 포스트를 불러오시겠습니까?
      </PopupOKCancel>
    </>
  );

  // return mode === WriteMode.MARKDOWN ? (
  //   <></>
  // ) : <QuillEditorContainer />
};

export default ActiveEditor;
