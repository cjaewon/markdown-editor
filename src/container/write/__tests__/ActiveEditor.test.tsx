import * as React from 'react';
import ActiveEditor, { ActiveEditorProps } from '../ActiveEditor';
import renderWithProviders from '../../../lib/renderWithProviders';
import { HelmetProvider } from 'react-helmet-async';
import { MemoryRouter } from 'react-router-dom';

describe('ActiveEditor', () => {
  const setup = (props: Partial<ActiveEditorProps> = {}) => {
    const utils = renderWithProviders(
      <MemoryRouter>
        <HelmetProvider>
          <ActiveEditor {...props} />
        </HelmetProvider>
      </MemoryRouter>,
    );
    return {
      ...utils,
    };
  };
  it('renders properly', () => {
    const utils = setup();
    utils.getByTestId('codemirror'); // initial mode: WYSIWYG
  });
  it('matches snapshot', () => {
    const { container } = setup();
    expect(container).toMatchSnapshot();
  });
  // it('converts editor properly', async () => {
  //   const { getByTestId, getByText } = setup({});
  //   const convertButton = getByTestId('quillconvert');
  //   fireEvent.click(convertButton);
  //   const confirmButton = getByText('확인');
  //   fireEvent.click(confirmButton);
  //   getByTestId('quill');
  // });
});
