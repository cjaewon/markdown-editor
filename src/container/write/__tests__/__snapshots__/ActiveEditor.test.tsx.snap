// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ActiveEditor matches snapshot 1`] = `
<div>
  <div
    class="sc-bdVaJa goJuwf"
  >
    <div
      class="sc-bwzfXH iBRLvj"
      data-testid="left"
      style="background-color: white;"
    >
      <div
        class="sc-cSHVUG ijAHOK"
        data-testid="codemirror"
      >
        <div
          class="wrapper"
        >
          <div
            style="max-height: 254px; opacity: 1;"
          >
            <div
              class="sc-chPdSV gpWtEG"
            >
              <textarea
                class="sc-ifAKCX dPCHAJ"
                placeholder="제목을 입력하세요"
                style="height: 0px;"
              />
              <div
                class="sc-kAzzGY eHnKgt"
              />
              <div
                class="sc-ckVGcZ fqvCvv"
              >
                <input
                  class="sc-jKJlTe ixuWhS"
                  placeholder="태그를 입력하세요"
                  tabindex="2"
                  value=""
                />
                <div
                  class="sc-dxgOiQ jWNXFC"
                />
              </div>
            </div>
          </div>
          <div
            class="sc-EHOje kKImpk"
            id="toolbar"
          >
            <button
              class="sc-htoDjs eGSEML ql-header"
              value="1"
            >
              <div
                class="sc-gzVnrw glwxBG"
              >
                H
                <span>
                  1
                </span>
              </div>
            </button>
            <button
              class="sc-htoDjs eGSEML ql-header"
              value="2"
            >
              <div
                class="sc-gzVnrw glwxBG"
              >
                H
                <span>
                  2
                </span>
              </div>
            </button>
            <button
              class="sc-htoDjs eGSEML ql-header"
              value="3"
            >
              <div
                class="sc-gzVnrw glwxBG"
              >
                H
                <span>
                  3
                </span>
              </div>
            </button>
            <button
              class="sc-htoDjs eGSEML ql-header"
              value="4"
            >
              <div
                class="sc-gzVnrw glwxBG"
              >
                H
                <span>
                  4
                </span>
              </div>
            </button>
            <div
              class="sc-dnqmqq ljEAXv"
            />
            <button
              class="sc-htoDjs eGSEML ql-bold"
            >
              <svg
                fill="currentColor"
                height="1em"
                stroke="currentColor"
                stroke-width="0"
                viewBox="0 0 24 24"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M15.6 10.79c.97-.67 1.65-1.77 1.65-2.79 0-2.26-1.75-4-4-4H7v14h7.04c2.09 0 3.71-1.7 3.71-3.79 0-1.52-.86-2.82-2.15-3.42zM10 6.5h3c.83 0 1.5.67 1.5 1.5s-.67 1.5-1.5 1.5h-3v-3zm3.5 9H10v-3h3.5c.83 0 1.5.67 1.5 1.5s-.67 1.5-1.5 1.5z"
                />
              </svg>
            </button>
            <button
              class="sc-htoDjs eGSEML ql-italic"
            >
              <svg
                fill="currentColor"
                height="1em"
                stroke="currentColor"
                stroke-width="0"
                viewBox="0 0 24 24"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M10 4v3h2.21l-3.42 8H6v3h8v-3h-2.21l3.42-8H18V4z"
                />
              </svg>
            </button>
            <button
              class="sc-htoDjs eGSEML ql-strike"
            >
              <svg
                fill="currentColor"
                height="1em"
                stroke="currentColor"
                stroke-width="0"
                viewBox="0 0 24 24"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M10 19h4v-3h-4v3zM5 4v3h5v3h4V7h5V4H5zM3 14h18v-2H3v2z"
                />
              </svg>
            </button>
            <div
              class="sc-dnqmqq ljEAXv"
            />
            <button
              class="sc-htoDjs eGSEML ql-blockquote"
            >
              <svg
                fill="currentColor"
                height="1em"
                stroke="currentColor"
                stroke-width="0"
                viewBox="0 0 24 24"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M6 17h3l2-4V7H5v6h3zm8 0h3l2-4V7h-6v6h3z"
                />
              </svg>
            </button>
            <button
              class="sc-htoDjs eGSEML ql-link"
            >
              <svg
                fill="currentColor"
                height="1em"
                stroke="currentColor"
                stroke-width="0"
                viewBox="0 0 24 24"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M3.9 12c0-1.71 1.39-3.1 3.1-3.1h4V7H7c-2.76 0-5 2.24-5 5s2.24 5 5 5h4v-1.9H7c-1.71 0-3.1-1.39-3.1-3.1zM8 13h8v-2H8v2zm9-6h-4v1.9h4c1.71 0 3.1 1.39 3.1 3.1s-1.39 3.1-3.1 3.1h-4V17h4c2.76 0 5-2.24 5-5s-2.24-5-5-5z"
                />
              </svg>
            </button>
            <button
              class="sc-htoDjs eGSEML ql-image"
            >
              <svg
                fill="currentColor"
                height="1em"
                stroke="currentColor"
                stroke-width="0"
                viewBox="0 0 24 24"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M21 19V5c0-1.1-.9-2-2-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2zM8.5 13.5l2.5 3.01L14.5 12l4.5 6H5l3.5-4.5z"
                />
              </svg>
            </button>
            <button
              class="sc-htoDjs eGSEML ql-code-block"
            >
              <svg
                fill="currentColor"
                height="1em"
                stroke="currentColor"
                stroke-width="0"
                viewBox="0 0 24 24"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M9.4 16.6L4.8 12l4.6-4.6L8 6l-6 6 6 6 1.4-1.4zm5.2 0l4.6-4.6-4.6-4.6L16 6l6 6-6 6-1.4-1.4z"
                />
              </svg>
            </button>
            <div
              class="mobile-placeholder"
            />
          </div>
          <div
            class="sc-kgoBCf bWKhLX"
          >
            <textarea
              style="display: none;"
            />
            <div
              class="CodeMirror cm-s-one-light CodeMirror-wrap CodeMirror-empty"
            >
              <div
                style="overflow: hidden; position: relative; width: 3px; height: 0px;"
              >
                <textarea
                  autocapitalize="off"
                  autocorrect="off"
                  spellcheck="false"
                  style="position: absolute; bottom: -1em; padding: 0px; width: 1000px; height: 1em; outline: none;"
                  tabindex="0"
                />
              </div>
              <div
                class="CodeMirror-vscrollbar"
                cm-not-content="true"
                tabindex="-1"
              >
                <div
                  style="min-width: 1px;"
                />
              </div>
              <div
                class="CodeMirror-hscrollbar"
                cm-not-content="true"
                tabindex="-1"
              >
                <div
                  style="height: 100%; min-height: 1px;"
                />
              </div>
              <div
                class="CodeMirror-scrollbar-filler"
                cm-not-content="true"
              />
              <div
                class="CodeMirror-gutter-filler"
                cm-not-content="true"
              />
              <div
                class="CodeMirror-scroll"
                tabindex="-1"
              >
                <div
                  class="CodeMirror-sizer"
                  style="margin-left: 0px;"
                >
                  <div
                    style="position: relative;"
                  >
                    <div
                      class="CodeMirror-lines"
                      role="presentation"
                    >
                      <div
                        role="presentation"
                        style="position: relative; outline: none;"
                      >
                        <pre
                          class="CodeMirror-placeholder CodeMirror-line-like"
                          style="height: 0px; overflow: visible; direction: ltr;"
                        >
                          당신의 이야기를 적어보세요...
                        </pre>
                        <div
                          class="CodeMirror-measure"
                        >
                          <pre
                            class="CodeMirror-line-like"
                          >
                            <span>
                              xxxxxxxxxx
                            </span>
                          </pre>
                        </div>
                        <div
                          class="CodeMirror-measure"
                        />
                        <div
                          style="position: relative; z-index: 1;"
                        />
                        <div
                          class="CodeMirror-cursors"
                        />
                        <div
                          class="CodeMirror-code"
                          role="presentation"
                        />
                      </div>
                    </div>
                  </div>
                </div>
                <div
                  style="position: absolute; height: 30px; width: 1px;"
                />
                <div
                  class="CodeMirror-gutters"
                  style="display: none;"
                />
              </div>
            </div>
          </div>
        </div>
        <div
          class="sc-kGXeez hcbBCR"
          style="width: 50%;"
        >
          <div
            class="sc-hMqMXs gLyvHa"
          >
            <button
              class="sc-iAyFgw ibDznP"
            >
              <svg
                fill="currentColor"
                height="1em"
                stroke="currentColor"
                stroke-width="0"
                viewBox="0 0 24 24"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M20 11H7.83l5.59-5.59L12 4l-8 8 8 8 1.41-1.41L7.83 13H20v-2z"
                />
              </svg>
              <span>
                나가기
              </span>
            </button>
            <div
              class="sc-kEYyzF eLvYji"
            >
              <button
                class="sc-fjdhpX jXoOWc sc-kkGfuU kNVwQV"
                color="lightGray"
              >
                임시저장
              </button>
              <button
                class="sc-fjdhpX ilmKPu sc-kkGfuU kNVwQV"
                color="teal"
              >
                출간하기
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="sc-bwzfXH sc-htpNat guBesd"
      data-testid="right"
      style="background-color: rgb(251, 253, 252);"
    >
      <div
        class="sc-brqgnP bNesev"
        id="preview"
      >
        <h1
          class="sc-cMljjf hxoyKh"
        />
        <div
          class="sc-cvbbAY cgBRCb"
        >
          <div
            class="sc-jWBwVP OSClR atom-one-light"
          >
            

          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
